[iOS] DOM nodes can be leaked when searching for text in Safari

pxlcoder opened the pull request:
#### c40691a5705e5b0d8ade87482cb7aef4bb86c385
<pre>
[iOS] DOM nodes can be leaked when searching for text in Safari
<a href="https://bugs.webkit.org/show_bug.cgi?id=285450">https://bugs.webkit.org/show_bug.cgi?id=285450</a>
<a href="https://rdar.apple.com/133689631">rdar://133689631</a>

Reviewed by Abrar Rahman Protyasha and Richard Robinson.

The UIKit API for find-in-page on iOS expects clients to vend all found ranges to
the system. The system then tells the client to highlight / scroll to the appropriate
range. Since converting between the range representation for UIKit and `SimpleRange`s
can be slow, a cache of found ranges is maintained.

Under most API use (via the system search UI), the cache is populated when the user
starts searching and is cleared when the user dismisses the find bar. However, when
searches are performed using Safari&apos;s URL bar, the find APIs may be called directly,
without the full system user flow that clears caches. Since the cache preserves
`SimpleRange`s, it strongly holds on to DOM nodes.

Fix by updating the cache to store `WeakSimpleRange`s. Additionally, clear the cache
on memory pressure.

* Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
(WebKit::WebFoundTextRangeController::findTextRangesForStringMatches):
(WebKit::WebFoundTextRangeController::clearAllDecoratedFoundText):
(WebKit::WebFoundTextRangeController::clearCachedRanges):
(WebKit::WebFoundTextRangeController::simpleRangeFromFoundTextRange):
* Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::releaseMemory):

Canonical link: <a href="https://commits.webkit.org/288505@main">https://commits.webkit.org/288505@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0e3bf6ae602d029fcde5961073ea1530320d0c7b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/83573 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/3190 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/37872 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/88644 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/34581 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/3278 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/11148 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/65016 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/22756 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/86619 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/2411 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/75936 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/45303 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/2318 "Passed tests") | [⏳ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-WK1-Tests-EWS "Waiting to run tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/33629 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/73395 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/30884 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/90021 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/10838 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/7825 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/73446 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/11061 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/71757 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/72669 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/16912 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/2162 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12906 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/10790 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/16262 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/10638 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/14113 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/12410 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
```diff
@@ -513,7 +518,7 @@ WebCore::Document* WebFoundTextRangeController::documentForFoundTextRange(const
 std::optional<WebCore::SimpleRange> WebFoundTextRangeController::simpleRangeFromFoundTextRange(WebFoundTextRange range)
 {
-    return m_cachedFoundRanges.ensure(range, [&] () -> std::optional<WebCore::SimpleRange> {
+    auto foundRange = m_cachedFoundRanges.ensure(range, [&] () -> std::optional<WebCore::WeakSimpleRange> {
```
Optional since this is pre-existing code but you can fix the style guide violation? i.e. no space between capture list and arg list.

In fact, the arg list can be omitted entirely now.

```suggestion
    auto foundRange = m_cachedFoundRanges.ensure(range, [&] -> std::optional<WebCore::WeakSimpleRange> {
```

pxlcoder committed:
[iOS] DOM nodes can be leaked when searching for text in Safari
https://bugs.webkit.org/show_bug.cgi?id=285450
rdar://133689631

Reviewed by Abrar Rahman Protyasha and Richard Robinson.

The UIKit API for find-in-page on iOS expects clients to vend all found ranges to
the system. The system then tells the client to highlight / scroll to the appropriate
range. Since converting between the range representation for UIKit and `SimpleRange`s
can be slow, a cache of found ranges is maintained.

Under most API use (via the system search UI), the cache is populated when the user
starts searching and is cleared when the user dismisses the find bar. However, when
searches are performed using Safari's URL bar, the find APIs may be called directly,
without the full system user flow that clears caches. Since the cache preserves
`SimpleRange`s, it strongly holds on to DOM nodes.

Fix by updating the cache to store `WeakSimpleRange`s. Additionally, clear the cache
on memory pressure.

* Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
(WebKit::WebFoundTextRangeController::findTextRangesForStringMatches):
(WebKit::WebFoundTextRangeController::clearAllDecoratedFoundText):
(WebKit::WebFoundTextRangeController::clearCachedRanges):
(WebKit::WebFoundTextRangeController::simpleRangeFromFoundTextRange):
* Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::releaseMemory):

Canonical link: https://commits.webkit.org/288505@main
